### PR TITLE
Support HTTP Range header for file-serving functionality.

### PR DESCRIPTION
**Pull request by fredrikekre**

### CODE REVIEW

File: src/resource.jl [hunk 1]
```diff
@@ -332,8 +332,30 @@ function serve_file(
...
 )
-    size = filesize(path)
-    HTTP.setheader(http, "Content-Length" => string(size))
+    content_length = filesize(path)
+    startbyte, stopbyte = 0, content_length-1
+
+    # Support single byte ranges
+    range_string = HTTP.header(http, "Range")
+    if !isempty(range_string)
+        m = match(r"^bytes=(\d*)-(\d*)$", range_string)
```
Comment (StefanKarpinski) [inline src/resource.jl:341]:
```suggestion
        m = match(r"^\s*bytes\s*=\s*(\d+)\s*-\s*(\d+)\s*$", range_string)
```

Comment (StefanKarpinski) [inline src/resource.jl]:
```suggestion
        m = match(r"^\s*bytes\s*=\s*(\d*)\s*-\s*(\d*)\s*$", range_string)
```
Let's be a bit more liberal with allowing spaces. Are the zero-repetition options here intentional? I.e. is the logic that the regex allows no digits, which fails to parse below but it replaced by the default? If so, maybe a comment clarifying that. That took me a while to work out and I had to look at the examples to be sure that it was intentional.

File: src/resource.jl [hunk 2]
```diff
@@ -332,8 +332,30 @@ function serve_file(
...
+        if m === nothing
+            @error "malformed HTTP Range header, ignoring" range_string
```
Comment (StefanKarpinski) [inline src/resource.jl]:
```suggestion
            @error "unsupported HTTP Range request, ignoring" range_string
```
The regex doesn't work for multipart range requests, which are valid but not supported, so this covers that too.

File: src/resource.jl [hunk 3]
```diff
@@ -344,13 +366,14 @@ function serve_file(
 
     # Open the path, write it out directly to the HTTP stream in chunks
     open(path) do io
+        seek(io, startbyte)
         t = 0
-        while !eof(io)
+        while t < content_length
             n = readbytes!(io, buffer)
```
Comment (StefanKarpinski) [inline src/resource.jl]:
```suggestion
            r = min(length(buffer), content_length - t)
            n = readbytes!(io, buffer, r)
```
This avoids reading the entire buffer when less data is needed.

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]

**Review by staticfloat — APPROVED** [review]
This is really cool!

**Comment by StefanKarpinski** [comment]
Yes. This paves the way for resuming incomplete downloads, which would be a really great feature in the Pkg client.

**Comment by fredrikekre** [comment]
> This paves the way for resuming incomplete downloads, which would be a really great feature in the Pkg client.

Yes, that's what I am implementing. Will do it as a local PkgServer for now.

File: src/resource.jl (hunk 1 above)
Comment (fredrikekre) [reply src/resource.jl]: > Let's be a bit more liberal with allowing spaces.

Sure, I didn't see spaces in any examples but I guess it doesn't hurt.

> Are the zero-repetition options here intentional? I.e. is the logic that the regex allows no digits, which fails to parse below but it replaced by the default?

Yea, thought it was a cute way of doing it but maybe it is clearer to just check for a match.

Comment (fredrikekre) [reply src/resource.jl]: Eh, I guess one should `tryparse` anyway for safety? I will just add a comment.

File: src/resource.jl (hunk 3 above)
Comment (fredrikekre) [reply src/resource.jl]: Yea. Does it work though? I just saw https://github.com/JuliaLang/julia/issues/36300. I will just leave it as a comment for now, can be optimized later.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by fredrikekre** [comment]
This failed setting up the test-server twice in a row on Travis, but works locally. Lets try again...

**Comment by fredrikekre** [comment]
Third time in a row: https://travis-ci.org/github/JuliaPackaging/PkgServer.jl/jobs/698815068#L279 but I see that it also happens on master sometimes.

**Comment by staticfloat** [comment]
I believe the timeout just needs to be increased.  I've got a fix in https://github.com/JuliaPackaging/PkgServer.jl/pull/52/commits/6ea7988612231f14d93b87991d0bc741a7f2a771

**Comment by fredrikekre** [comment]
Okay, I cherry-picked that commit here.

**Comment by StefanKarpinski** [comment]
@staticfloat, I'll let you merge when you're ready since you're juggling other changes to this.

**Pull request merged**